core/src/main/java/com/linecorp/armeria/server/AbstractBindingBuilder.java

```diff
@@ -0,0 +1,279 @@
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.MediaType;
+
+/**
+ * An abstract builder class binding something to a {@link Route} fluently.
+ */
+abstract class AbstractBindingBuilder {
+
+    private Set<HttpMethod> methods = HttpMethod.knownMethods();
+    private Set<MediaType> consumeTypes = ImmutableSet.of();
+    private Set<MediaType> produceTypes = ImmutableSet.of();
+    private final Map<RouteBuilder, Set<HttpMethod>> routeBuilders = new LinkedHashMap<>();
+    private final ImmutableSet.Builder<RouteBuilder> pathBuilders = ImmutableSet.builder();
+
+    /**
+     * Sets the path pattern that a {@link Service} will be bound to.
+     *
+     * @throws IllegalArgumentException if the specified path pattern is invalid
+     */
+    public AbstractBindingBuilder path(String pathPattern) {
+        pathBuilders.add(Route.builder().path(requireNonNull(pathPattern, "pathPattern")));
+        return this;
+    }
+
+    /**
+     * Sets the specified prefix which is a directory that a {@link Service} will be bound under.
+     *
+     * @throws IllegalArgumentException if the specified path pattern is invalid
+     */
+    public AbstractBindingBuilder pathUnder(String prefix) {
+        pathBuilders.add(Route.builder().prefix(requireNonNull(prefix, "prefix")));
+        return this;
+    }
+
+    /**
+     * Sets the path pattern that a {@link Service} will be bound to, only supporting {@link HttpMethod#GET}
+     * requests.
+     *
+     * @throws IllegalArgumentException if the specified path pattern is invalid
+     */
+    public AbstractBindingBuilder get(String pathPattern) {
+        addRouteBuilder(pathPattern, GET);
+        return this;
+    }
+
+    /**
+     * Sets the path pattern that a {@link Service} will be bound to, only supporting {@link HttpMethod#POST}
+     * requests.
+     *
+     * @throws IllegalArgumentException if the specified path pattern is invalid
+     */
+    public AbstractBindingBuilder post(String pathPattern) {
+        addRouteBuilder(pathPattern, POST);
+        return this;
+    }
+
+    /**
+     * Sets the path pattern that a {@link Service} will be bound to, only supporting {@link HttpMethod#PUT}
+     * requests.
+     *
+     * @throws IllegalArgumentException if the specified path pattern is invalid
+     */
+    public AbstractBindingBuilder put(String pathPattern) {
+        addRouteBuilder(pathPattern, PUT);
+        return this;
+    }
+
+    /**
+     * Sets the path pattern that a {@link Service} will be bound to, only supporting {@link HttpMethod#PATCH}
+     * requests.
+     *
+     * @throws IllegalArgumentException if the specified path pattern is invalid
+     */
+    public AbstractBindingBuilder patch(String pathPattern) {
+        addRouteBuilder(pathPattern, PATCH);
+        return this;
+    }
+
+    /**
+     * Sets the path pattern that a {@link Service} will be bound to, only supporting {@link HttpMethod#DELETE}
+     * requests.
+     *
```

Comment:
How about listing the example patterns here, just like we did for `ServerBuilder.service(String, Service)`?